vimiv/imutils/exif.py

```diff
@@ -261,13 +293,95 @@ def copy_exif(self, dest: str, reset_orientation: bool = True) -> None:
         except OSError as e:
             _logger.debug("Failed to write exif data for '%s': '%s'", dest, str(e))
 
-    def exif_date_time(self) -> str:
+    def get_date_time(self) -> str:
         with contextlib.suppress(KeyError):
             return self._metadata["Exif.Image.DateTime"].raw_value
         return ""
 
 
-has_exif_support = ExifHandler != _ExifHandlerBase
+has_exif_support = _ExternalKeyHandler != _ExternalKeyHandlerBase
+
+
+class MetadataHandler:
+    """Handler to load and copy exif information of a single image.
+
+    This class provides the interface for handling exif support. By default none of the
+    operations are implemented. Instead it is up to a child class which wraps around one
+    of the supported exif libraries to implement the methods it can.
+    """
+
+    def __init__(self, filename=""):
+        self.filename = filename
+        self._internal_handler: Optional[_InternalKeyHandler] = None
+        self._external_handler: Optional[_ExternalKeyHandler] = None
+
+    @property
+    def internal_handler(self) -> _InternalKeyHandler:
+        if self._internal_handler is None:
+            self._internal_handler = _InternalKeyHandler(self.filename)
+        return self._internal_handler
+
+    @property
+    def external_handler(self) -> _ExternalKeyHandler:
+        if self._external_handler is None:
+            self._external_handler = _ExternalKeyHandler(self.filename)
+        return self._external_handler
+
+    def copy_metadata(self, dest: str, reset_orientation: bool = True) -> None:
+        """Copy exif information from current image to dest.
+
+        Args:
+            dest: Path to write the exif information to.
+            reset_orientation: If true, reset the exif orientation tag to normal.
+        """
+        try:
+            self.external_handler.copy(dest, reset_orientation)
+        except UnsupportedExifOperation:
+            # Todo
+            pass
+
+    def get_date_time(self) -> str:
+        """Get exif creation date and time as formatted string."""
+        try:
+            self.external_handler.get_data_time()
+        except UnsupportedExifOperation:
+            # Todo
+            pass
+
+    def get_formatted_metadata(
+        self, desired_keys: Sequence[str]
+    ) -> Dict[Any, Tuple[str, str]]:
+        metadata = dict()
+
+        for base_key in desired_keys:
+            base_key = base_key.strip()
+
+            try:
+                key, key_name, key_value = self._fetch_key(base_key)
+                metadata[key] = key_name, key_value
+            except (KeyError, TypeError):
+                _logger.debug("Invalid key '%s'", base_key)
+
+        return metadata
+
+    def get_keys(self) -> Iterable[str]:
+        """Retrieve the name of all exif keys available."""
+        try:
+            return itertools.chain(
+                self.internal_handler.get_keys(), self.external_handler.get_keys()
+            )
+        except UnsupportedExifOperation:
```

Comment:
I guess we would have to return an empty list or so. But I am also fine with letting the exception propagate, like this the caller can directly decide what to do when there is no exif support.

Reply:
I think it makes more sense to just propagate the exception. I have done that in 31cddcb942705a911d07823a6ab013bc41fddfb3. Additionallty, I have added a notice to the docstrings that clients must fetch the UnsuportedExifOperation exception.